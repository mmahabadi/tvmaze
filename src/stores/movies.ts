import { defineStore } from 'pinia'
import { computed, ref, type ComputedRef } from 'vue'
import * as ApiService from '@/services/ApiService'
import { type Movie, type ShowByGenre } from '@/types'

export const useMoviesStore = defineStore('movies', () => {
  const shows = ref<Map<string, ShowByGenre>>(new Map())
  const movies = ref<Movie[]>([])
  const details = ref<Movie | null>(null)
  const loading = ref(false)
  const error = ref<null | Error>(null)
  const currentIndex = ref(1)
  const ITEMS_PER_PAGE = 3
  const ITEMS_LIMIT_FOR_RECENT_MOVIES = 15
  const DATE_LIMIT_FOR_RECENT_MOVIES = 2020

  const recentMovies = computed(() =>
    movies.value
      .filter((movie) => new Date(movie.ended).getFullYear() > DATE_LIMIT_FOR_RECENT_MOVIES)
      .slice(0, ITEMS_LIMIT_FOR_RECENT_MOVIES)
  )

  /**
   * @note because the API does not support fetching genres,
   * I need to extract the genres from the fetched movies
   */
  const genres = computed(() => [...new Set([...movies.value.flatMap((movie) => movie.genres)])])

  const getByGenre = computed(() => (genre: string): ShowByGenre => {
    loadGenre(genre)
    return shows.value.get(genre) as ShowByGenre
  })

  const paginatedGenres: ComputedRef<string[]> = computed(() => {
    const startIndex = currentIndex.value
    const endIndex = startIndex * ITEMS_PER_PAGE
    // Concatenate the new slice with the previous values of pagedGenres
    return [...genres.value.slice(0, endIndex)]
  })

  async function fetchMovies(): Promise<void> {
    loading.value = true
    try {
      movies.value = await ApiService.getData()
    } catch (e) {
      error.value = e as Error
    } finally {
      loading.value = false
    }
  }

  async function fetchShowByGenre(genre: string): Promise<void> {
    const genreShows = shows.value.get(genre)
    if (!genreShows) return

    try {
      /**
       * @note if the APi supports fetching movies by genre,
       * comment the following line
       * and uncomment the line below
       */
      const items = movies.value
        .filter((movie) => movie.genres.includes(genre))
        .sort((a, b) => b.rating.average - a.rating.average)
      /**
       * @note the API does not support fetching movies by genre
       * uncomment the following line to fetch the movies by genre from the API
       */
      // const items = await ApiService.getShowsByGenre(genre)
      if (items) {
        genreShows.movies = items
      }
    } catch (e) {
      genreShows.error = e as Error
    } finally {
      genreShows.loading = false
    }
    shows.value.set(genre, genreShows as ShowByGenre)
  }

  function loadGenre(genre: string): void {
    if (!shows.value.has(genre)) {
      const genreShows: ShowByGenre = {
        genre,
        movies: [],
        loading: true,
        error: null
      }
      shows.value.set(genre, genreShows)
      fetchShowByGenre(genre)
    }
  }

  function nextPage(): void {
    if (currentIndex.value * ITEMS_PER_PAGE <= genres.value.length) currentIndex.value++
  }

  async function getShowDetails(id: number): Promise<void> {
    loading.value = true
    try {
      details.value = await ApiService.getDetails(id)
    } catch (e) {
      error.value = e as Error
    } finally {
      loading.value = false
    }
  }

  return {
    movies,
    genres: paginatedGenres,
    recentMovies,
    details,
    loading,
    error,
    fetchMovies,
    getByGenre,
    nextPage,
    getShowDetails
  }
})
