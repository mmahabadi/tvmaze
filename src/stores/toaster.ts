import { defineStore } from 'pinia'
import { ref } from 'vue'
import type { ToasterMessage } from '@/types'

export const useToasterStore = defineStore('toaster', () => {
  const toasters = ref<ToasterMessage[]>([])
  let idCounter = 0

  const addToaster = (message: string, type: 'warning' | 'error' | 'success') => {
    toasters.value.push({ id: idCounter++, message, type })
  }

  const removeToaster = (id: number) => {
    toasters.value = toasters.value.filter((toaster) => toaster.id !== id)
  }

  return {
    toasters,
    addToaster,
    removeToaster
  }
})
