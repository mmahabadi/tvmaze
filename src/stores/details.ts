import { defineStore } from 'pinia'
import { ref } from 'vue'
import * as ApiService from '@/services/ApiService'
import type { Movie } from '@/types'
import { useToasterStore } from './toaster'

export const useDetailsStore = defineStore('details', () => {
  const details = ref<Movie | null>(null)
  const loading = ref(false)
  const error = ref<null | Error>(null)
  const { addToaster } = useToasterStore()

  async function getShowDetails(id: number): Promise<void> {
    loading.value = true
    try {
      details.value = await ApiService.getDetails(id)
    } catch (e) {
      error.value = e as Error
      addToaster(`Failed to fetch the show`, 'error')
    } finally {
      loading.value = false
    }
  }

  return {
    details,
    loading,
    error,
    getShowDetails
  }
})
