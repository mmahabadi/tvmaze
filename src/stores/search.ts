import type { Movie } from '@/types'
import { defineStore } from 'pinia'
import { ref } from 'vue'
import * as ApiService from '@/services/ApiService'
import { useToasterStore } from './toaster'

export const useSearchStore = defineStore('search', () => {
  const searchResults = ref<Movie[]>([])
  let searchQuery = ''
  const loading = ref(false)
  const error = ref<null | Error>(null)

  const { addToaster } = useToasterStore()

  async function searchShows(query: string): Promise<void> {
    //avoid making the same request multiple times
    if (searchQuery === query) return
    searchQuery = query
    loading.value = true
    try {
      searchResults.value = await ApiService.search(query)
    } catch (e) {
      error.value = e as Error
      addToaster(`Failed to fetch shows`, 'error')
    } finally {
      loading.value = false
    }
  }

  return {
    searchResults,
    searchQuery,
    loading,
    error,
    searchShows
  }
})
