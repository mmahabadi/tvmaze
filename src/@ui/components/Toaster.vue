<script setup lang="ts">
import { ref, onMounted, onUnmounted, inject } from 'vue'
import SuccessIcon from '@/@ui/components/icons/SuccessIcon.vue'
import ErrorIcon from '@/@ui/components/icons/ErrorIcon.vue'
import WarningIcon from '@/@ui/components/icons/WarningIcon.vue'
import CloseIcon from '@/@ui/components/icons/CloseIcon.vue'
import { useToasterStore } from '@/stores/toaster'
import { defineProps } from 'vue'
import type { ToasterMessage, AppSettings } from '@/types'

const props = defineProps<ToasterMessage>()
const appSettings = inject<AppSettings>('appSettings')
const store = useToasterStore()

const timeout = ref<ReturnType<typeof setTimeout>>()

onMounted(() => {
  timeout.value = setTimeout(() => {
    onClose()
  }, appSettings?.toasterTimeout || 5000)
})

onUnmounted(() => {
  clearTimeout(timeout.value)
})

const onClose = () => {
  store.removeToaster(props.id)
}
</script>
<template>
  <div class="toaster" role="alert">
    <div :class="'icon ' + type">
      <SuccessIcon v-if="type === 'success'" />
      <ErrorIcon v-if="type === 'error'" />
      <WarningIcon v-if="type === 'warning'" />
    </div>
    <div class="text">{{ message }}</div>
    <button type="button" @click="onClose" class="close-button" aria-label="Close">
      <CloseIcon />
    </button>
  </div>
</template>
<style scoped>
.toaster {
  @apply flex items-center w-full max-w-xs p-4 mb-4 rounded-lg shadow text-gray-400 bg-gray-800;

  .icon {
    @apply inline-flex items-center justify-center flex-shrink-0 w-8 h-8 rounded-lg;

    &.success {
      @apply bg-green-800 text-green-200;
    }
    &.warning {
      @apply bg-orange-700 text-orange-200;
    }
    &.error {
      @apply bg-red-800 text-red-200;
    }
  }
  .text {
    @apply ms-3 text-sm font-normal;
  }

  .close-button {
    @apply ms-auto -mx-1.5 -my-1.5 rounded-lg focus:ring-2 focus:ring-gray-300 p-1.5 inline-flex items-center justify-center h-8 w-8 text-gray-500 hover:text-white bg-gray-800 hover:bg-gray-700;
  }
}
</style>
