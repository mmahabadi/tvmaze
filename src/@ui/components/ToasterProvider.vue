<script setup lang="ts">
import Toaster from './Toaster.vue'
import { useToasterStore } from '@/stores/toaster'

const store = useToasterStore()
</script>
<template>
  <div class="fixed z-40 top-32 right-5">
    <Toaster
      v-for="toaster in store.toasters"
      :key="toaster.id"
      :message="toaster.message"
      :type="toaster.type"
      :id="toaster.id"
    />
  </div>
</template>
