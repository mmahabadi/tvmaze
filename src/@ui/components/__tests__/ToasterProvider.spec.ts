import { mount, type VueWrapper } from '@vue/test-utils'
import { beforeEach, describe, expect, it, vi } from 'vitest'
import ToasterProvider from '../ToasterProvider.vue'
import { useToasterStore } from '@/stores/toaster'
import Toaster from '../Toaster.vue'

vi.mock('@/stores/toaster', () => {
  return {
    useToasterStore: vi.fn()
  }
})

const mockToasters = [
  { id: 1, message: 'Test Message 1', type: 'success' },
  { id: 2, message: 'Test Message 2', type: 'error' }
]

describe('ToasterProvider', () => {
  let wrapper: VueWrapper<unknown>

  beforeEach(() => {
    ;(useToasterStore as any).mockReturnValue({
      toasters: mockToasters
    })

    // setActivePinia(createPinia())
    wrapper = mount(ToasterProvider)
  })

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders the correct number of toasters', () => {
    const toasters = wrapper.findAllComponents(Toaster)
    expect(toasters).toHaveLength(mockToasters.length)
  })
})
