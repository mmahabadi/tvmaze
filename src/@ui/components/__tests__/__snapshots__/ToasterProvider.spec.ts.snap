// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ToasterProvider > renders correctly 1`] = `
"<div class="fixed z-40 top-32 right-5">
  <div data-v-a2f0cfc0="" class="toaster" role="alert">
    <div data-v-a2f0cfc0="" class="icon success"><svg data-v-a2f0cfc0="" class="w-5 h-5" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" fill="currentColor" viewBox="0 0 20 20">
        <path d="M10 .5a9.5 9.5 0 1 0 9.5 9.5A9.51 9.51 0 0 0 10 .5Zm3.707 8.207-4 4a1 1 0 0 1-1.414 0l-2-2a1 1 0 0 1 1.414-1.414L9 10.586l3.293-3.293a1 1 0 0 1 1.414 1.414Z"></path>
      </svg>
      <!--v-if-->
      <!--v-if-->
    </div>
    <div data-v-a2f0cfc0="" class="text">Test Message 1</div><button data-v-a2f0cfc0="" type="button" class="close-button" aria-label="Close"><svg data-v-a2f0cfc0="" class="w-3 h-3" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 14 14">
        <path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6"></path>
      </svg></button>
  </div>
  <div data-v-a2f0cfc0="" class="toaster" role="alert">
    <div data-v-a2f0cfc0="" class="icon error">
      <!--v-if--><svg data-v-a2f0cfc0="" class="w-5 h-5" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" fill="currentColor" viewBox="0 0 20 20">
        <path d="M10 .5a9.5 9.5 0 1 0 9.5 9.5A9.51 9.51 0 0 0 10 .5Zm3.707 11.793a1 1 0 1 1-1.414 1.414L10 11.414l-2.293 2.293a1 1 0 0 1-1.414-1.414L8.586 10 6.293 7.707a1 1 0 0 1 1.414-1.414L10 8.586l2.293-2.293a1 1 0 0 1 1.414 1.414L11.414 10l2.293 2.293Z"></path>
      </svg>
      <!--v-if-->
    </div>
    <div data-v-a2f0cfc0="" class="text">Test Message 2</div><button data-v-a2f0cfc0="" type="button" class="close-button" aria-label="Close"><svg data-v-a2f0cfc0="" class="w-3 h-3" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 14 14">
        <path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6"></path>
      </svg></button>
  </div>
</div>"
`;
